### PR TITLE
Type id refactor

### PR DESCRIPTION
**Pull request by Lehdari**
Refactor type id system to provide compile time ids via constexpr interface in separate class.
The new method utilizes macros constructed by CMake but voids the need for component template macros and dummy source files.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,40 +1,62 @@
...
+#Set required C++ standard
 set(CMAKE_CXX_STANDARD 17)
 set(CMAKE_CXX_STANDARD_REQUIRED ON)
 
+#Output libraries to same directory with binaries
 set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR})
 set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR})
 
-add_subdirectory(ext)
-add_subdirectory(breakout)
```
Comment (tkln) [inline CMakeLists.txt:11]: Name change of the breakout directory should be in the previous commit, which is currently broken.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,40 +1,62 @@
...
+    # Find and link SFML
+    find_package(SFML 2.5 COMPONENTS system window graphics REQUIRED)
+
+    add_executable(Breakout
+        ${BREAKOUT_SOURCES}
+    )
+
+    target_include_directories(Breakout
+        PUBLIC
+            breakout/include
```
Comment (tkln) [inline CMakeLists.txt:47]: This should be Breakout/include.

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,40 +1,62 @@
...
+            fug_engine
+            vecmat
+            sfml-system
+            sfml-window
+            sfml-graphics
+    )
+
+    target_compile_definitions(Breakout
+        PRIVATE
+            RES_DIRECTORY="${CMAKE_CURRENT_SOURCE_DIR}/breakout"
```
Comment (tkln) [inline CMakeLists.txt:63]: This should have been changed in the previous commit.

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,40 +1,62 @@
...
-# Find and link SFML
-find_package(SFML 2.5 COMPONENTS system window graphics REQUIRED)
-
-add_executable(Breakout 
-    ${BREAKOUT_SOURCES}
-)
-
-target_include_directories(Breakout
-    PUBLIC
-        breakout/include
```
Comment (tkln) [inline CMakeLists.txt:22]: Same as above.

File: CMakeLists.txt (hunk 1 above)
Comment (tkln) [reply CMakeLists.txt:11]: Also CMake copies the source tree directory structure into the build folder and on Unix-like systems executables have no file extension. These two factors create a situtaion where we have build/Breakout directory while linker tries (and fails) to create an executable with the exact same name.

**Review by tkln — CHANGES_REQUESTED** [review]

**Comment by Lehdari** [comment]
Changes included into other PR

**Pull request closed without merging**